include/IDriverFence.h

```diff
@@ -27,7 +27,25 @@ enum E_DRIVER_FENCE_RETVAL
     EDFR_ALREADY_SIGNALED
 };
 
-//! Persistently Mapped buffer
+//! Base class for Fences, persistently Mapped buffer
+/*
+    Using fences is neccesarry while requesting operation for
+    GPU depended on CPU side, like fetching mapped GPU buffer.
+    It's significant because CPU and GPU execute processes asynchronously,
+    so before the CPU does anything for instance to mapped memory the GPU
+    has been writing to, waiting on a fence is needed and then if the mapped memory
+    needs to be invalidated before CPU reads it (to invalidate the CPU cache) you should
+    invalidate it by canDeferredFlush()

```

Comment:
`canDeferredFlush` does not invalidate the cache, `canDeferredFlush` is a property of `IDriverFence` and it tells you whether the `waitCPU` method can accept a `flush=true` parameter (which performs an implicit `glFlush` on the context you've placed the fence on just before you actually start witing on the fence).

A pipeline/context flush (`glFlush`) flushes the driver-CPU-side queue onto the GPU device for execution, if you place a fence but don't flush you can end up in a deadlock, basically each time you call `waitCPU` it will time-out because the work has not been sent to the GPU for execution 100% (its stuck in transit).

So you either need to perform a `glFlush` (but on the same thread+context that placed the fence) after placing the fence but before `waitCPU` (on any thread) or use that implicit flush functionality (but only if setting thread is the waiting thread), see the documentation for `waitCPU` .

Reply:
the correct function to use are these:
https://github.com/buildaworldnet/IrrlichtBAW/blob/shader_pipeline/include/IDriver.h#L147

its flush when CPU is writing and GPU is reading
invalidate when CPU is reading and GPU is writing

if you're not confident controlling cache coherency manually then use the COHERENT memory type when creating mappable `IGPUBuffer`s

Reply:
adjusted in following commit